### PR TITLE
Add spatial functionality to projections

### PR DESCRIPTION
**Pull request by e-koch**
Adds much of the spatial functionality from `SpectralCube` for a 2D spatial `Projection`.

* Added `SpatialCoordMixinClass` for world coordinates handling.
* Beam handling/loading for `Projection`
* `Projection.convolve_to` and `Projection.reproject`
* `Projection.subimage` for slicing with quantities (as in `SpectralCube.subcube`)
* `WCSCelestialError` is raised if the `Projection` does not have 2 spatial axes
*  Added 2D test images

Not all of the spatial functionality has been abstracted out of `SpectralCube`. Things like `SpectralCube.flattened_world` require mask handling to be implemented for lower dimensional objects. Same goes for `SpectralCube.subcube_from_ds9_region`, in its current implementation.

### CODE REVIEW

File: spectral_cube/spectral_cube.py [hunk 1]
```diff
@@ -81,7 +80,7 @@ def wrapper(*args, **kwargs):
 # conventions between WCS and numpy
 np2wcs = {2: 0, 1: 1, 0: 2}
 
-class BaseSpectralCube(BaseNDClass, SpectralAxisMixinClass):
+class BaseSpectralCube(BaseNDClass, SpectralAxisMixinClass, SpatialCoordMixinClass):
```
Comment (keflavich) [inline spectral_cube/spectral_cube.py]: I just learned recently... I think `super` always refers to the *last* of these 3, and so maybe we want `BaseNDClass` last?  But we should check with someone who knows better (...or rtfd...) before changing it

File: spectral_cube/spectral_cube.py [hunk 2]
```diff
@@ -183,20 +182,13 @@ def _new_cube_with(self, data=None, wcs=None, mask=None, meta=None,
...
 
-    def _try_load_beam(self, header):
+    def _attach_beam(self):
 
-        try:
-            from radio_beam import Beam
-        except ImportError:
-            warnings.warn("radio_beam is not installed. No beam "
-                          "can be created.")
+            beam = cube_utils.try_load_beam(self.header)
```
Comment (keflavich) [inline spectral_cube/spectral_cube.py]: is this now one level too far indented?

File: spectral_cube/base_class.py [hunk 3]
```diff
@@ -31,6 +38,120 @@ def meta(self):
...
+        coordinates, strong distortions)
+
+        This method is not currently implemented. Use :meth:`world` instead.
+        """
+        raise NotImplementedError()
+
+    @property
+    def spatial_coordinate_map(self):
+
+        if self.ndim > 2:
```
Comment (keflavich) [inline spectral_cube/base_class.py]: this is actually the case `if self.ndim == 3`.  Could do a general one:
```
view = [slice(0) for ii in range(self.ndim-2)] + [slice(None)]*2
return self.world[view][self.ndim-2:]
```

File: spectral_cube/base_class.py [hunk 4]
```diff
@@ -31,6 +38,120 @@ def meta(self):
...
+        """
+        Returns a list of 1D arrays, for the world coordinates
+        along each pixel axis.
+
+        Raises error if this operation is ill-posed (e.g. rotated world
+        coordinates, strong distortions)
+
+        This method is not currently implemented. Use :meth:`world` instead.
+        """
+        raise NotImplementedError()
```
Comment (keflavich) [inline spectral_cube/base_class.py]: This might be significantly more possible now if we can check `is_separable` on the wcs...

**Comment by keflavich** [comment]
Also, I don't know what exactly went wrong with the docs, but something went very wrong.

File: spectral_cube/lower_dimensional_structures.py [hunk 5]
```diff
@@ -246,6 +267,149 @@ def _quicklook_mpl(self, filename=None):
...
+
+        self._raise_wcs_no_celestial()
+
+        try:
+            from reproject.version import version
+        except ImportError:
+            raise ImportError("Requires the reproject package to be"
+                              " installed.")
+
+        # Need version > 0.2 to work with cubes
```
Comment (e-koch) [inline spectral_cube/lower_dimensional_structures.py]: I think so. We could just remove it here, or keep it so the requirements are homogeneous.

File: spectral_cube/base_class.py [hunk 6]
```diff
@@ -1,16 +1,23 @@
...
 DOPPLER_CONVENTIONS = {}
 DOPPLER_CONVENTIONS['radio'] = u.doppler_radio
 DOPPLER_CONVENTIONS['optical'] = u.doppler_optical
 DOPPLER_CONVENTIONS['relativistic'] = u.doppler_relativistic
 
 
-
 class BaseNDClass(object):
+
+    _cache = {}
```
Comment (e-koch) [inline spectral_cube/base_class.py]: Needed for caching. `BaseSpectralCube` does this as well: https://github.com/radio-astro-tools/spectral-cube/blob/master/spectral_cube/spectral_cube.py#L133.

Switching to `lazyproperty` seems like the better route. Maybe that should be separate PR?

Comment (keflavich) [inline spectral_cube/base_class.py]: Yes, separate PR is a good idea.

**Comment by e-koch** [comment]
@keflavich -- The docs issue is from the LDOs inheriting from `Quantity`. For some reason, `doctest_skip` isn't ignoring the docstrings from `ndarray`. Any thoughts?

The docs *do* build fine, at least locally, with all of these warnings.

**Comment by keflavich** [comment]
No immediate ideas, sorry.  Maybe someone in astropy will have a clue...

File: spectral_cube/lower_dimensional_structures.py [hunk 7]
```diff
@@ -163,10 +165,11 @@ def quantity(self):
         """
         return u.Quantity(self)
 
-class Projection(LowerDimensionalObject):
+class Projection(LowerDimensionalObject, SpatialCoordMixinClass):
```
Comment (keflavich) [inline spectral_cube/lower_dimensional_structures.py]: What if you swapped the order of the inheritance?  Maybe that would fix the doc issues?

**Comment by keflavich** [comment]
As @e-koch noted, this is related to https://github.com/astropy/astropy/pull/2159.  I wonder if there's a way to add numpy docs to the namespace, then the inherited docs wouldn't be a problem?  Otherwise, maybe we change something about docstring inheritance...?

File: docs/nitpick-exceptions [hunk 8]
```diff
@@ -9,3 +9,35 @@ py:class spectral_cube.base_class.SpatialCoordMixinClass
 py:obj yt.surface.export_sketchfab
 py:obj yt.show_colormaps
 
+# numpy inherited docstrings
```
Comment (e-koch) [inline docs/nitpick-exceptions]: 👍  for brute force

**Comment by keflavich** [comment]
Everything passes now.  Good to merge @e-koch?
